### PR TITLE
[Lang] Add ti.Vector.ndarray and ti.Matrix.ndarray

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #2771

This PR adds `ti.Vector.ndarray` and `ti.Matrix.ndarray`, which support Python-scope and Taichi-scope access. When passing them into kernels, we need to use `ti.any_arr()` annotation with `element_shape` and optional `layout` (default value: `ti.Layout.AOS`) to enforce type consistency.
```python
@ti.kernel
def func(a: ti.any_arr(element_shape=(2, 2), layout=ti.Layout.SOA)):
    for i in range(5):
        for j, k in ti.ndrange(2, 2):
            a[i][j, k] = j * j + k * k

m = ti.Matrix.ndarray(2, 2, ti.i32, 5, layout=ti.Layout.SOA)
func(m)
```

### CODE REVIEW

File: python/taichi/lang/ndarray.py [hunk 1]
```diff
@@ -84,11 +81,32 @@ def shape(self):
...
 
     def __repr__(self):
         return '<ti.ndarray>'
+
+
+class NdarrayHostAccess:
+    """Class for accessing VectorNdarray/MatrixNdarray in Python scope.
+    Args:
+        arr (Union[VectorNdarray, MatrixNdarray]): See above.
+        indices_first (Tuple[Int]): Indices of first-level access.
```
Comment (k-ye) [inline python/taichi/lang/ndarray.py:98]: What are first level and second level?

File: python/taichi/lang/kernel_arguments.py [hunk 2]
```diff
@@ -57,8 +58,25 @@ def __init__(self, element_shape=(), layout=Layout.AOS):
         self.layout = layout
 
     def extract(self, x):
-        shape = tuple(x.shape)
+        from taichi.lang.matrix import MatrixNdarray, VectorNdarray
```
Comment (k-ye) [inline python/taichi/lang/kernel_arguments.py:61]: move to top? (or will it cause circular import?)

Comment (strongoier) [reply python/taichi/lang/kernel_arguments.py:61]: It will cause circular import :-(

File: python/taichi/lang/ndarray.py (hunk 1 above)
Comment (strongoier) [reply python/taichi/lang/ndarray.py:98]: First level refers to coordinates in the field, while second level refers to indices inside the vector/matrix. Maybe I should make it clearer?

Comment (strongoier) [reply python/taichi/lang/ndarray.py:98]: Using first/second is to tell users this is what they write in the first/second pair of brackets.

File: tests/python/test_ndarray.py [hunk 3]
```diff
@@ -29,3 +111,95 @@ def run(x: ti.any_arr(), y: ti.any_arr()):
...
+    v = ti.Vector.ndarray(10, ti.i32, 5, layout=layout)
+    func(v)
+    assert v[0][6] == 0
+    assert v[1][0] == 0
+    assert v[2][1] == 1
+    assert v[3][4] == 4
+    assert v[4][9] == 9
+
+
+# number of compiled functions
```
Comment (ailzhang) [inline tests/python/test_ndarray.py:184]: is this comment not longer needed?

File: python/taichi/lang/matrix.py [hunk 4]
```diff
@@ -1435,3 +1482,96 @@ def empty(self):
...
+    def __getitem__(self, key):
+        key = () if key is None else (
+            key, ) if isinstance(key, numbers.Number) else tuple(key)
+        return Matrix.with_entries(self.n, self.m, [
+            NdarrayHostAccess(self, key, (i, j)) for i in range(self.n)
+            for j in range(self.m)
+        ])
+
+    def __repr__(self):
+        return f'<{self.n}x{self.m} ti.Matrix.ndarray>'
```
Comment (ailzhang) [inline python/taichi/lang/matrix.py:1535]: IMHO printing out layout information will also help?

File: python/taichi/lang/matrix.py [hunk 5]
```diff
@@ -1435,3 +1482,96 @@ def empty(self):
...
+
+    @python_scope
+    def __getitem__(self, key):
+        key = () if key is None else (
+            key, ) if isinstance(key, numbers.Number) else tuple(key)
+        return Matrix.with_entries(
+            self.n, 1,
+            [NdarrayHostAccess(self, key, (i, )) for i in range(self.n)])
+
+    def __repr__(self):
```
Comment (ailzhang) [inline python/taichi/lang/matrix.py:1576]: ditto

**Review by ailzhang — COMMENTED** [review]
LGTM!

File: tests/python/test_ndarray.py (hunk 3 above)
Comment (strongoier) [reply tests/python/test_ndarray.py:184]: It's used to separate three different parts of tests of this file.

File: python/taichi/lang/matrix.py (hunk 4 above)
Comment (strongoier) [reply python/taichi/lang/matrix.py]: Nice suggestion. Done.

File: python/taichi/lang/matrix.py (hunk 5 above)
Comment (strongoier) [reply python/taichi/lang/matrix.py:1576]: done

File: python/taichi/lang/ndarray.py (hunk 1 above)
Comment (strongoier) [reply python/taichi/lang/ndarray.py]: Explanations are added.

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request merged**